torch/_dynamo/bytecode_analysis.py

```diff
@@ -164,55 +165,55 @@ def walk(state, start):
     return must.reads | may.reads
 
 
-@dataclasses.dataclass
-class FixedPointBox:
-    value: bool = True
-
-
 @dataclasses.dataclass
 class StackSize:
     low: Real
     high: Real
-    fixed_point: FixedPointBox
 
     def zero(self):
         self.low = 0
         self.high = 0
-        self.fixed_point.value = False
 
     def offset_of(self, other, n):
         prior = (self.low, self.high)
         self.low = min(self.low, other.low + n)
         self.high = max(self.high, other.high + n)
         if (self.low, self.high) != prior:
-            self.fixed_point.value = False
+            return True
+        else:
+            return False
 
 
 def stacksize_analysis(instructions):
     assert instructions
-    fixed_point = FixedPointBox()
+
     stack_sizes = {
-        inst: StackSize(float("inf"), float("-inf"), fixed_point)
-        for inst in instructions
+        inst: StackSize(float("inf"), float("-inf")) for inst in instructions
     }
     stack_sizes[instructions[0]].zero()
 
-    for _ in range(100):
-        if fixed_point.value:
-            break
-        fixed_point.value = True
-
-        for inst, next_inst in zip(instructions, instructions[1:] + [None]):
-            stack_size = stack_sizes[inst]
-            if inst.opcode not in TERMINAL_OPCODES:
-                assert next_inst is not None, f"missing next inst: {inst}"
-                stack_sizes[next_inst].offset_of(
-                    stack_size, stack_effect(inst.opcode, inst.arg, jump=False)
-                )
-            if inst.opcode in JUMP_OPCODES:
-                stack_sizes[inst.target].offset_of(
-                    stack_size, stack_effect(inst.opcode, inst.arg, jump=True)
-                )
+    indexof = {id(inst): i for i, inst in enumerate(instructions)}
+
+    worklist = deque()
+    worklist.append(0)
+
+    while len(worklist) != 0:
+        index = worklist.popleft()
+        inst = instructions[index]
+        stack_size = stack_sizes[inst]
+        if inst.opcode not in TERMINAL_OPCODES:
+            assert index + 1 < len(instructions), f"missing next inst: {inst}"
+            changed = stack_sizes[instructions[index + 1]].offset_of(
+                stack_size, stack_effect(inst.opcode, inst.arg, jump=False)
+            )
+            if changed:
+                worklist.append(index + 1)
+        if inst.opcode in JUMP_OPCODES:
+            changed = stack_sizes[inst.target].offset_of(
+                stack_size, stack_effect(inst.opcode, inst.arg, jump=True)
+            )
+            if changed:
```

Comment:
[Here](https://github.com/pytorch/pytorch/issues/97811#issuecomment-1591888590) is a repro which causes your branch to loop infinitely (with `python3.8`).

This is an existing bug (#97811) which I was trying to fix (#103621), but it seems easier to address here.

The issue is that we are not correctly modelling the `CALL_FINALLY` opcode - see the `cpython3.8` implementation [here](https://github.com/python/cpython/blob/3.8/Python/compile.c#L5499).

I think this patch should solve the issue:

```suggestion
            if changed and (
                sys.version_info >= (3, 9) or inst.opcode != dis.opmap["CALL_FINALLY"]
            ):
```

Reply:
Sorry for the late reply! Thanks a lot for your suggestion! I will check this later. Maybe this PR can be merged after the fix.

Reply:
Sounds good - in the meantime, #103621 did go in, so rebasing might have fixed it. That being said, I think doing it this way lines up better with the CPython implementation.